lib/src/main/java/com/auth0/jwt/oicmsg/KeyJar.java

```diff
@@ -0,0 +1,377 @@
+                    continue;
+                }
+                if(key.getUse() != null || use.equals(key.getUse())) {
+                    if(kid != null) {
+                        if(key.getKid().equals(kid)) {
+                            keyListReturned.add(key);
+                            break;
+                        }
+                    } else {
+                        keyListReturned.add(key);
+                    }
+                }
+            }
+        }
+
+        String name;
+        if(keyType.equals("EC") && args.containsKey("alg")) {
+            name = "P-{}" + args.get("alg").substring(2);
+            List<Key> tempKeyList = new ArrayList<>();
+            for(Key key : keyList) {
+                try {
+                    Assert.assertTrue(name.equals(((ECKey) key).getCrv()));
+                } catch (AssertionError error) {
+                    continue;
+                } finally {
+                    tempKeyList.add(key);
+                }
+            }
+            keyList = tempKeyList;
+        }
+
+        if(use.equals("enc") && keyType.equals("oct") && owner != null && !owner.isEmpty()) {
+            for(KeyBundle keyBundle : this.issuerKeys.get("")) {
+                for(Key key : keyBundle.get(keyType)) {
+                    if(key.getUse() == null || key.getUse() == use) {
+                        keyList.add(key);
+                    }
+                }
+            }
+        }
+
+        return keyList;
+    }
+
+    public List<Key> getSigningKey(String keyType, String owner, String kid, Map<String,String> args) {
+        return getKeys("sig", keyType, owner, kid, args);
+    }
+
+    public List<Key> getVerifyKey(String keyType, String owner, String kid, Map<String,String> args) {
+        return getKeys("ver", keyType, owner, kid, args);
+    }
+
+    public List<Key> getEncryptKey(String keyType, String owner, String kid, Map<String,String> args) {
+        return getKeys("enc", keyType, owner, kid, args);
+    }
+
+    public List<Key> getDecryptKey(String keyType, String owner, String kid, Map<String,String> args) {
+        return getKeys("dec", keyType, owner, kid, args);
+    }
+
+    public List<Key> keysByAlgAndUsage(String issuer, String algorithm, String usage) {
+        String keyType;
+        if(usage.equals("sig") || usage.equals("ver")) {
+            keyType = algorithmToKeytypeForJWS(algorithm);
+        } else {
+            keyType = algorithmToKeytypeForJWE(algorithm);
+        }
+
+        return getKeys(usage, keyType, issuer, null, null);
+    }
+
+    public List<Key> getIssuerKeys(String issuer) {
+        List<Key> keyList = new ArrayList<>();
+        for(KeyBundle keyBundle : this.issuerKeys.get(issuer)) {
+            keyList.addAll(keyBundle.getKeys());
+        }
+        return keyList;
+    }
+
+    private String algorithmToKeytypeForJWS(String algorithm) {
+        if(algorithm == null || algorithm.toLowerCase().equals("none")) {
```

Comment:
are you sure that `algorithm == null` is going to be "none" instead of throwing some exception?

Reply:
That's how it's done in Roland's code.  You have to remember I'm porting this over from the python version of oicmsg to java

Reply:
I'll clarify my comment as it looks cryptic. When the algorithm is "none" there's no doubt the key would be "none". But if the algorithm is null, do you really want this to be considered as "none" as well, without any warning?

Reply:
that's how it's done in Roland's code.  i guess we should stick to the same implementation.

Reply:
It's common in python to regard None and '' as similar and representing no value.
If that is not the case in Java then you should probably treat None and '' differently.

Reply:
what do you mean by "?  empty string?

i thought None and null were the same.